lang/cs/Org.Apache.REEF.Examples/DriverRestart/HelloRestartDriver.cs

```diff
@@ -0,0 +1,181 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+using System;
+using System.Diagnostics;
+using System.Text;
+using Org.Apache.REEF.Common.Tasks;
+using Org.Apache.REEF.Driver;
+using Org.Apache.REEF.Driver.Context;
+using Org.Apache.REEF.Driver.Evaluator;
+using Org.Apache.REEF.Driver.Task;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Util;
+using Org.Apache.REEF.Utilities.Logging;
+using IRunningTask = Org.Apache.REEF.Driver.Task.IRunningTask;
+
+namespace Org.Apache.REEF.Examples.DriverRestart
+{
+    /// <summary>
+    /// The Driver for HelloRestartREEF.
+    /// This driver is meant to run on YARN on HDInsight, with the ability to keep containers
+    /// across application attempts.
+    /// It requests 5 evaluators and runs a running task on each of the evaluators.
+    /// Once all tasks are running, the driver kills itself and expects the RM to restart it.
+    /// On restart, it expects all 5 of the running tasks to report back to it.
+    /// </summary>
+    public sealed class HelloRestartDriver : IObserver<IDriverRestartCompleted>, IObserver<IAllocatedEvaluator>, IObserver<IDriverStarted>, 
+        IObserver<IDriverRestarted>, IObserver<IActiveContext>, IObserver<IRunningTask>, IObserver<ICompletedTask>, IObserver<IFailedTask>
+    {
+        private static readonly Logger _Logger = Logger.GetLogger(typeof(HelloRestartDriver));
+        private const int AmountOfTasksToSubmit = 5;
+
+        private readonly IEvaluatorRequestor _evaluatorRequestor;
+        private readonly object _lockObj;
+
+        private int _runningTaskCount;
+        private int _finishedTaskCount;
+        private bool _restarted;
+
+        [Inject]
+        private HelloRestartDriver(IEvaluatorRequestor evaluatorRequestor)
+        {
+            _finishedTaskCount = 0;
+            _runningTaskCount = 0;
+            _evaluatorRequestor = evaluatorRequestor;
+            _restarted = false;
+            _lockObj = new object();
+        }
+
+        /// <summary>
+        /// This constructor is for the configuration to know that the driver has a restart
+        /// handler bound for it to enable restart, and is not intended for actual instantiation.
+        /// </summary>
+        [Inject]
+        private HelloRestartDriver()
```

Comment:
It's needed in ``ClrClient2JavaClientCuratedParameters`` to check if there are RestartHandles bound to enable restart. Just having the constructor with ``IEvaluatorRequestor`` would fail because that particular injector is not able to inject ``IEvaluatorRequestor``. I'm not sure if there's a better way around this because I would like to avoid adding a boolean parameter just to enable restart on the .NET side.